### PR TITLE
ci: add QG2 platform readiness gate

### PR DESCRIPTION
**Pull request by kami619**
## Description

- Add a reusable `run-qg2` composite action and standalone `QG2: Platform Readiness` workflow
- Keep QG2 strict and profile-aware for operator health, `DataScienceCluster`, and KServe platform readiness
- Align the initial implementation to the live `rhoai2` cluster reality and register QG2 in the CI dashboard and alerting docs

## Jira Ticket

- RHAIENG-5866
- RHAIENG-5827

## Testing

- [ ] `make test` passes (run from the affected agent directory)
- [x] Manual testing performed (describe steps below)
- [ ] No testing required (documentation/config change only)

Verification:
- `uv run pytest .github/scripts/tests -q` (`109 passed`)
- `uv run ruff check .github/scripts/qg2_platform_readiness.py .github/scripts/tests/test_qg2_platform_readiness.py .github/scripts/tests/test_qg2_workflow_contract.py .github/scripts/generate_ci_health_page.py .github/scripts/tests/test_generate_ci_health_page.py`
- `uv run ruff format --check .github/scripts/qg2_platform_readiness.py .github/scripts/tests/test_qg2_platform_readiness.py .github/scripts/tests/test_qg2_workflow_contract.py .github/scripts/generate_ci_health_page.py .github/scripts/tests/test_generate_ci_health_page.py`
- Python YAML parse sanity check for the new action/workflows
- Live cluster inspection on `rhoai2` for operator, `DataScienceCluster`, and KServe controller signals

## Checklist

- [x] I have read [CONTRIBUTING.md](CONTRIBUTING.md)
- [x] No `.env` or secret files are included in this PR
- [x] All changes are within scope of the linked Jira ticket (if not, explain in Description)

## Review Guidance

- QG2 is intentionally a singleton pre-matrix gate, mirroring the QG1 delivery shape.
- `setup-cluster` remains the shared login/bootstrap layer; `run-qg2` owns only platform readiness assertions.
- Current target assumption: align to the live `rhoai2` cluster, not generic `odh` behavior.
- AIA message: Implemented with Cursor in an isolated worktree using a subagent-driven flow, with final acceptance review, local verification, and live-cluster inspection before push.

AI-Attribution: AIA PAI Nc Hin R gpt-5.4 v1.0
AI-Interpretation: https://aiattribution.github.io/statements/AIA-PAI-Nc-Hin-R-?model=gpt-5.4

### CODE REVIEW

**Review by shricharan-ks — DISMISSED** [review]
## Multi-Perspective Review — PR #318

Reviewed from four independent angles: **Python correctness**, **security**, **test quality**, and **design/scope**.

---

### Verdict: APPROVE

The QG2 platform readiness gate is well-engineered. The checker script has careful error discrimination (`run_oc_optional` for NotFound vs real errors), proper operator fallback logic (deployment → CSV), and clean separation between evaluation functions and `oc` interaction. The workflow mirrors QG1's structure closely, and the 740 lines of tests are thorough. All CI checks pass.

---

### What this PR does well

- **Error discrimination is excellent** — `run_oc_optional` distinguishes genuine NotFound from RBAC/timeout errors, preventing the gate from silently misreporting a Forbidden error as "operator absent". This is the hardest part of writing a cluster probe, and it's done correctly.
- **CSV fallback only applies when deployment is genuinely absent** — a present-but-unhealthy deployment is NOT rescued by a stale "Succeeded" CSV phase. This prevents the "everything looks green" trap.
- **No command injection paths** — all `subprocess.run` calls use list form (no `shell=True`), inputs are either hardcoded constants or argparse-validated choices, and the composite action passes inputs via `env:` blocks (with a test enforcing this).
- **Strong QG1/QG2 consistency** — identical checkout pins, same setup-cluster → resolve → run → upload → logout flow, same notify-slack job shape.
- **Tests cover the tricky edge cases** — operator absent with CSV fallback, DSC absent vs. DSC not ready, deployment exists but unhealthy even with Succeeded CSV, and real oc errors (Forbidden) propagating correctly through the generic error path.

---

### Suggestions

#### 1. [Security — Medium] QG2 should get a dedicated service account

If PR #319 merges first, QG2 becomes the only quality gate using the raw bootstrap `OC_TOKEN` directly. QG2 needs broader cluster-wide permissions (CSVs, DataScienceCluster, deployments with `-A`) than QG1. A dedicated `qg2-readiness` SA with scoped RBAC would limit blast radius — or better, generalize the `assume-qg1-service-account` action from PR #319 into a shared `assume-service-account` action that both gates can use. The runbook correctly documents the raw-token requirement, but recommend tracking this as a follow-up (or coordinating with RHAIENG-6993).

#### 2. [Test gap] Missing integration tests for disabled checks through `main()`

`--require-dsc-ready false` and `--require-kserve false` are only tested at the `evaluate_*` unit level. An integration test through `main()` would verify the flag correctly short-circuits the `oc` calls (fewer `run_oc` invocations), not just the evaluation logic. This matters because the `values = iter([...])` pattern would silently break if the call count changes.

#### 3. [Test gap] No test for `subprocess.TimeoutExpired`

`main()` catches `TimeoutExpired` and `build_error_summary` handles it with a specific branch, but no test exercises this path. A monkeypatched test raising `TimeoutExpired` would confirm the summary file is written and exit code is 1.

#### 4. [Test gap] Missing step-ordering assertion in workflow contract tests

`test_qg2_workflow_uses_shared_setup_and_qg2_action` asserts both actions exist but doesn't verify `setup-cluster` runs before `run-qg2`. The QG1 contract tests include an ordering assertion (`setup_idx < assume_idx < run_idx`). QG2 should match:

```python
setup_idx = uses_values.index("./.github/actions/setup-cluster")
run_idx = uses_values.index("./.github/actions/run-qg2")
assert setup_idx < run_idx
```

#### 5. [Test quality] `test_fixture_loads_all_workflows` lost its field assertions

The count bump from 5→6 is correct, but the original test also verified `summaries[0].display_name` and `summaries[0].latest.conclusion`. These assertions were dropped rather than moved. The new `test_qg2_fixture_entry_is_included` covers QG2 specifically but the original QG1/Code Quality field assertions are now uncovered.

#### 6. [Minor] Non-UTF-8 `oc` output would crash without writing summary files

`subprocess.run(..., text=True)` raises unhandled `UnicodeDecodeError` if `oc` emits non-UTF-8 bytes. This isn't caught by the `CalledProcessError`/`TimeoutExpired` handler, so the gate would crash without writing results. Unlikely in practice, but `encoding="utf-8", errors="replace"` would make it robust.

---

### Not a concern

- **PR size**: The 1555 additions break down as checker (386) + tests (914) + action/workflow (197) + docs/dashboard (58). The bulk is test coverage, and the components are tightly coupled — splitting would create merge-ordering dependencies for minimal review benefit.
- **Schedule timing**: QG1 at 01:00 UTC, QG2 at 01:15 UTC. Both are read-only probes on independent resource types. 15-minute gap is sufficient; no shared mutable state.
- **TOCTOU race on `.items[0]`**: A resource deleted between the names query and the `.items[0]` query would produce a slightly imprecise detail message ("phase: missing" vs "not found") but the verdict (fail) would still be correct.
- **`_to_int("")` returning 0**: Correct — a deployment without `readyReplicas` status is genuinely not ready.

LGTM — thorough implementation with good error handling and strong test coverage.

**Comment by kami619** [comment]
Addressed items 2-6 in `0d64a69`.

- Disabled `--require-dsc-ready` / `--require-kserve` now short-circuit the corresponding `oc` probes in `main()`, so disabled gates no longer fail on unnecessary reads.
- Added direct coverage for the `TimeoutExpired` summary path.
- Restored the setup-before-run workflow contract assertion.
- Restored representative fixture field assertions in the CI health dashboard test.
- Hardened `oc` subprocess decoding with `encoding="utf-8", errors="replace"`.

Item 1 is reasonable follow-up work, but I left it out of this PR because the dedicated service-account shape overlaps the broader auth/generalization work referenced in PR #319 / RHAIENG-6993.

Verified with:
- `uv run pytest .github/scripts/tests/test_qg2_platform_readiness.py`
- `uv run pytest .github/scripts/tests/test_qg2_workflow_contract.py`
- `uv run pytest .github/scripts/tests/test_generate_ci_health_page.py`

AI-Attribution: AIA PAI Ce Hin R gpt-5.4 v1.0
AI-Interpretation: https://aiattribution.github.io/statements/AIA-PAI-Ce-Hin-R-?model=gpt-5.4

File: docs/ci-alert-policy.md [hunk 1]
```diff
@@ -73,7 +74,7 @@ flowchart TD
   QG9 -->|"retry from QG4"| QG4
 ```
 
-Only `QG1`, `QG4`, and `QG7` emit alerts today. The remaining fingerprints are
+Only `QG1`, `QG2`, `QG4`, and `QG7` emit alerts today. The remaining fingerprints are
```
[the remark this answers is not in the document]
Comment (kami619) [reply docs/ci-alert-policy.md:77]: don't worry about this, there is a future docs revision effort coming, these are superficial and would get corrected once we implement that new docs view.

File: .github/workflows/qg2-platform-readiness.yml [hunk 2]
```diff
@@ -0,0 +1,127 @@
...
+
+jobs:
+  qg2:
+    name: "QG2 Platform Readiness"
+    if: github.repository == 'red-hat-data-services/agentic-starter-kits'
+    runs-on: ubuntu-latest
+    timeout-minutes: 15
+    steps:
+      - name: Checkout
+        uses: actions/checkout@3d3c42e5aac5ba805825da76410c181273ba90b1 # v7.0.1
```
[the remark this answers is not in the document]
Comment (kami619) [reply .github/workflows/qg2-platform-readiness.yml:37]: The persist-credentials: false suggestion on the main QG2 checkout is weak. In qg2-platform-readiness.yml, the main qg2 job checks out code and then runs local actions and oc; it is not doing git writes. More importantly, qg1-cluster-readiness.yml uses the same pattern in its main job, and this PR was intentionally keeping QG1/QG2 aligned. I’d only take that change as part of a broader workflow-hardening sweep, not as a one-off “fix” here.

File: .github/scripts/qg2_platform_readiness.py [hunk 3]
```diff
@@ -0,0 +1,394 @@
...
+                )
+                kserve_desired_replicas = run_oc(
+                    "get",
+                    "deployment",
+                    "-A",
+                    "-l",
+                    KSERVE_LABEL_SELECTOR,
+                    "-o",
+                    "jsonpath={.items[0].spec.replicas}",
+                )
```
[the remark this answers is not in the document]
Comment (kami619) [reply .github/scripts/qg2_platform_readiness.py:361]: The “fetch once as JSON and parse in Python” comment is partly right, but it’s more of a refactor than a bug fix. In qg2_platform_readiness.py, the current pattern intentionally uses targeted oc jsonpath calls plus explicit absent-resource checks. Moving to one -o json fetch per resource family would reduce call count and could make tests less order-coupled, but it also changes the failure surface and only makes the JSONDecodeError advice relevant if you adopt that refactor. I’d treat this as optional cleanup in a later iteration.

File: .github/scripts/qg2_platform_readiness.py [hunk 4]
```diff
@@ -0,0 +1,394 @@
...
+    parser.add_argument("--cluster-profile", required=True)
+    parser.add_argument("--cluster-type", choices=["odh", "rhoai"], default="rhoai")
+    parser.add_argument("--require-dsc-ready", default="true")
+    parser.add_argument("--require-kserve", default="true")
+    parser.add_argument("--summary-json", required=True)
+    parser.add_argument("--summary-md", required=True)
+    args = parser.parse_args(argv)
+
+    require_dsc_ready = args.require_dsc_ready.lower() == "true"
+    require_kserve = args.require_kserve.lower() == "true"
```
[the remark this answers is not in the document]
Comment (kami619) [reply .github/scripts/qg2_platform_readiness.py:247]: The realistic failure cases are:

  • a future workflow edit that passes a mistyped string
  • a bad repo variable value like QG2_REQUIRE_DSC_READY=yes
  • someone reusing run-qg2 from another workflow incorrectly
  • someone running qg2_platform_readiness.py directly with a typo

I am willing to accept the risk for these as reviews would likely catch them., I am leaning towards keeping this as is.

**Comment by kami619** [comment]
@shricharan-ks can you please give another approval, I have implemented the changes you suggested, all good ones :)

**Review by shricharan-ks — APPROVED** [review]
LGTM
